### PR TITLE
fix(providers): Handle Azure AD tenants correctly

### LINKED ISSUE #9635: AzureAD common tenant URLs aren't being checked properly
### Provider type

Azure Active Directory

### Environment

System:
OS: macOS 14.2.1
CPU: (12) arm64 Apple M2 Max
Memory: 65.83 MB / 32.00 GB
Shell: 5.9 - /bin/zsh
Binaries:
Node: 21.2.0 - ~/.nvm/versions/node/v21.2.0/bin/node
npm: 10.2.3 - ~/.nvm/versions/node/v21.2.0/bin/npm
pnpm: 8.12.0 - ~/Library/pnpm/pnpm
bun: 1.0.14 - ~/.bun/bin/bun
Browsers:
Brave Browser: 120.1.61.109
Safari: 17.2.1
npmPackages:
@auth/core: ^0.18.6 => 0.18.6
@auth/sveltekit: ^0.4.0 => 0.4.0

### Reproduction URL

https://github.com/JibbityJobbity/next-auth-example/

### Describe the issue

Azure's implementation of OAuth requires manipulating/replacing parts of the URL based on your tenant ID which you are using. Next-auth's codebase frequently has the returned authentication/other endpoints checked and raises an error otherwise. This happens on any app with the common tenant ID, meaning just a generic Microsoft login like you would expect with just about any other OAuth provider. But you can have a look at how I'm using it in my project.

There are two checks where I've noticed that it's an issue.

The location the first check fails is in `getAuthorizationUrl` in `@auth/core/lib/actions/signin/authorization-url.js`, where we have this check `o.processDiscoveryResponse` expect the response url to be `https://login.microsoftonline.com/common/v2.0` when what we get back in actuality is `https://login.microsoftonline.com/{tenantid}/v2.0`. Microsoft expects users to replace `{tenantid}` with `common`, or whatever other tenant ID that's supplied when the AzureAD provider is set up.
For this, @panva has kindly provided a code snippet:

```js
import * as oauth from 'oauth4webapi'
import * as jose from 'jose'

const as = await oauth
  .discoveryRequest(new URL('https://login.microsoftonline.com/common/v2.0'))
  .then((response) =>
    oauth.processDiscoveryResponse(
      new URL('https://login.microsoftonline.com/{tenantid}/v2.0'),
      response,
    ),
  )

const response = oauth.authorizationCodeGrantRequest(as, client, params, redirectUri, codeVerifier)

let tenantId
if (response.status === 200) {
  const { id_token } = await response.clone().json()
  ;({ tid: tenantId } = jose.decodeJwt(id_token))
}

if (!isThisTenantIdExpected(tenantId)) {
  throw new Error()
}

const result = await oauth.processAuthorizationCodeOpenIDResponse(
  {
    ...as,
    issuer: as.issuer.replace('{tenantid}', tenantId),
  },
  client,
  response,
)
```

The second one is in `handleOauth` in where `o.processAuthorizationCodeOpenIDResponse` gets called. The way it eventually makes its way down will get the user's specific "issuer" tenant in the returned url. We tell the check to expect 'https://login.microsoftonline.com/common/v2.0' when the returned URL in the oauth4web library will get the same URL from Microsoft, but have `common` replaced with some tenant ID, I assume it's the user's.

I'd make a PR but I have no clue how to fit a solution in that's compatible with the project's architecture.

### How to reproduce

Use the AzureAD provider with the "common" tenant ID and log in.

### Expected behavior

Login to work with Microsoft's special endpoints with the tenant IDs that it expects. The checks drill themselves all the way down to oauth4web, but we kinda tell it to check for the wrong thing.

### PR DESCRIPTION
**Pull request by JibbityJobbity**
## ☕️ Reasoning

Endpoints returned by Azure AD want us to edit the path so that each request gets routed to their proper tenant IDs.
The old implementation didn't handle this properly when using the "common" tenant ID, which basically lets us use it normally like any other provider. We would receive the template endpoint paths and use those in error checking and stuff rather than the ones we actually use.

I've just thrown something up and I've done some quick testing. It seems to work normally. I'll test custom tenants at some point and I suspect it won't work until I test and fix it.

## 🧢 Checklist

- [X] Documentation
- [X] Tests
- [X] Ready to be merged

## 🎫 Affected issues

Fixes: #9635

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by JibbityJobbity** [comment]
Testing needed

- Common tenant group
- organizations and consumers tenant groups
- Any other specific tenant ID
- Other providers still work

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Any non-compliance needs to be handled at the provider level. we don't want branching based on specific providers, it's not maintainable in the long term

This will need a different solution. Something similar to the `conform` method on `token`.


some useful info at the end of the thread here https://github.com/nextauthjs/next-auth/issues/8831

**Comment by panva** [comment]
I think there should be two distinct Azure providers

One where you know the tenant you're working with and you're NOT using any of the "common" issuers (the existing one with the exception that tenant id is required).

And then another one which can be used with one of the "common" issuers

```
  'https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration'
  'https://login.microsoftonline.com/organizations/v2.0/.well-known/openid-configuration'
  'https://login.microsoftonline.com/consumers/v2.0/.well-known/openid-configuration'
```

This one has the workarounds in and requires explicit configuration to

- allow login from all tenants
- allow login from specific tenants (by providing a list of tenant ids), the ID Token's `tid`  would be checked to match the expected tenants.

**Comment by JibbityJobbity** [comment]
While we all know the real solution to this, I think by the time I have a crack at this, we will know what ends up happening. But that's definitely considerable if `conform()` doesn't end up happening.
However in the meantime, I think the `conform()` method is the first way forward with this to try. It provides a way for us to sort out LinkedIn too, but I'd be careful about getting too comfortable with hacking around providers as I can imagine `conform()` would make it easier to do exactly that.

File: packages/core/src/providers/azure-ad.ts
Comment (panva) [inline packages/core/src/providers/azure-ad.ts:144]: This doesn't do anything, does it? The Response `url` property is readonly and immutable

File: packages/core/src/providers/azure-ad.ts
Comment (panva) [inline packages/core/src/providers/azure-ad.ts:154]: same

File: packages/core/src/providers/azure-ad.ts
Comment (JibbityJobbity) [reply packages/core/src/providers/azure-ad.ts:144]: You're probably right. I'm not a TS/JS dev but I should've picked something like this up.
I'd hold off on review until I say something.

File: packages/core/src/lib/actions/callback/oauth/callback.ts
Comment (JibbityJobbity) [inline packages/core/src/lib/actions/callback/oauth/callback.ts:141]: I'm having some trouble here. It might be the way my development environment is set up, but it seems to think `serverConform` isn't a part of the object.

**Comment by JibbityJobbity** [comment]
I've done what I can for now. Completely open to suggestions and feedback.

I know this doesn't quite match the sample, but I wanted the check to work the other way around. Rather than having it try to check against the "fake" URL with `{tenantid}`, I thought it would be cleaner to perform a real check.

File: packages/core/src/providers/azure-ad.ts
Comment (JibbityJobbity) [inline packages/core/src/providers/azure-ad.ts:156]: This isn't right. Need to detect the tenant that's being used. Probably just replace the provider's tenant ID with whatever we expect from the JWT.

**Comment by panva** [comment]
I'm not finding all the individual bits from my [original recommendation](https://github.com/panva/oauth4webapi/issues/100#issuecomment-1890380232) nor the recommended split from https://github.com/nextauthjs/next-auth/pull/9718#issuecomment-1909893145 so I'm assuming you're going with your own workarounds.

**Comment by JibbityJobbity** [comment]
> I'm not finding all the individual bits from my [original recommendation](https://github.com/panva/oauth4webapi/issues/100#issuecomment-1890380232) nor the recommended split from [#9718 (comment)](https://github.com/nextauthjs/next-auth/pull/9718#issuecomment-1909893145) so I'm assuming you're going with your own workarounds.

It's a mix between what you suggested in the other issue and the conform methods. It still achieves the same thing, which is satisfying those underlying checks in your library while bringing the conformity to the provider level, as suggested by @balazsorban44. If this isn't good enough then I'll change the PR to satisfy the solution you gave earlier. The special `serverConform` is a bit icky so I see where you're coming from.

Anyway, I just did some testing with the other grouped tenants. The issuer URL comes back with some weird tenant ID instead of getting the user to replace "{tenantid}" or some other value we can expect. Replacing "{tenantid}" *only* seems to work with "common" and doesn't work with "organizations" or "consumers".  This keeps rearing its ugly head, you almost need something to the effect of this to override the check to bypass it:

```js
    const discoveryResponse: Response | undefined = await o.discoveryRequest(issuer)
    let json = await response.json();
    const as = await o.processDiscoveryResponse(new URL(json.issuer), discoveryResponse!)
```

I'll need another go at this. I'll shoot for tomorrow.

**Comment by JibbityJobbity** [comment]
This seems to work nicely :) I'm largely happy with the current solution now.

**Comment by JibbityJobbity** [comment]
Bump, rebased.

**Comment by amplicity** [comment]
hey! great work here, we could really use this. would love to see this get merged soon.

**Comment by MrLoh** [comment]
It would be great to see this get merged. I'm sure it blocks a lot of people from adopting v5.

**Comment by jaysin586** [comment]
Is there an update on this?

**Comment by ladparth** [comment]
Hello,
Thank you for working on this PR. I am also stuck on this stage of my development. Is there any update on this fix?

**Comment by JibbityJobbity** [comment]
> Hello, Thank you for working on this PR. I am also stuck on this stage of my development. Is there any update on this fix or is there any temporary fix i can do while this PR is in process?

I'll rebase, but I haven't heard anything from any of the maintainers since I got it working.
Do keep in mind they may be busy with their lives, but there are some other hacks to make it work by configuring the provider itself. This PR would just mean that nobody would have to be concerned about this again.

**Comment by ladparth** [comment]
Thank you @JibbityJobbity

**Comment by murilobd** [comment]
Thanks @JibbityJobbity for that PR! Hopefully maintainers of this repo will merge it!

**Comment by Ghernouz** [comment]
Any updates guys?

**Comment by RonB** [comment]
It looks like there is some status quo here. Azure AD also has been rebranded to Microsoft Entra Id and that has been adressed in the code already. I am afraid the train is moving on and a solution is getting out of site. @JibbityJobbity, thank you for your hard work. A request on behalf of all of us if you could adress the remaining open ends here?

**Comment by MrLoh** [comment]
Entra ID has the exact same issue that this PR tries to fix.

**Comment by JibbityJobbity** [comment]
I've rebased and put the fix onto Entra ID. I haven't changed any logic.
Seeing as people have shown interest in getting this fixed, I'd like to kindly ask the community to try testing this out for me. I'm going to have to pull the "life stuff" excuse for not getting this done sooner or testing it.
Hopefully this gets merged once some people say it works.

**Comment by RonB** [comment]
@JibbityJobbity 
Thank you so much for rebasing, I will advocate for this change to be merged but I am having a little trouble getting it to work.
I have a mono-repo with 2 apps and use pnpm. How should i use your repo?
In my package.json something like:

```json
"dependencies":
    "@auth/core": "https://github.com/JibbityJobbity/next-auth.git",
    "@auth/sveltekit": "^1.1.0"
"pnpm": {
    "overrides": {
      "@auth/core": "https://github.com/JibbityJobbity/next-auth.git"
    }
  },
  ```
  
 It errors with: 
 ```
 Error: Failed to resolve entry for package "@auth/core". The package may have incorrect main/module/exports specified in its package.json.
 ```
 Does the package has to be build first and do i need to make a fork of your repo and do that myself or is there a simpler solution use your repo?
 I am not an expert on this depency management stuff so i hope you can point me in the right direction.
 
 Excuse my ignorance :-)
 Thank you for your response

**Comment by michal-boruczkowski** [comment]
@RonB
I was trying to use it like this:
```
dependencies: {
     ...
    "next-auth": "git://github.com/JibbityJobbity/next-auth.git#azure-ad-common-tenant",
}
```

but I'm getting following error when `npm i`:
```
npm error git dep preparation failed
npm error npm warn using --force Recommended protections disabled.
npm error npm error code EUNSUPPORTEDPROTOCOL
npm error npm error Unsupported URL Type "workspace:": workspace:*
```

I'm using `patch` prepared here: https://github.com/nextauthjs/next-auth/issues/8374#issuecomment-1803387771 ATM

Would be great if this fix is merged 🤞

**Comment by JibbityJobbity** [comment]
Yeah, the way I got around it for development was to hack stuff in my `node_modules` folder. Not entirely sure how to deploy it in my own web app actually :/ if I use the sveltekit framework package in my branch then it tries to pull the core package from npm instead of my branch. Everyone I ask about this says "idk publish the package on npm". Sorry I don't have any useful info.

**Comment by hiromijorge** [comment]
> @RonB I was trying to use it like this:
> 
> ```
> dependencies: {
>      ...
>     "next-auth": "git://github.com/JibbityJobbity/next-auth.git#azure-ad-common-tenant",
> }
> ```
> 
> but I'm getting following error when `npm i`:
> 
> ```
> npm error git dep preparation failed
> npm error npm warn using --force Recommended protections disabled.
> npm error npm error code EUNSUPPORTEDPROTOCOL
> npm error npm error Unsupported URL Type "workspace:": workspace:*
> ```
> 
> `pnpm` seems to handle this but actual fix is not applied in runtime
> 
> I'm using `patch` prepared here: [#8374 (comment)](https://github.com/nextauthjs/next-auth/issues/8374#issuecomment-1803387771) ATM
> 
> Would be great if this fix is merged 🤞

Hi Michael,

I've followed the instructions in this [link](https://github.com/nextauthjs/next-auth/issues/8374#issuecomment-1803387771) for applying the patch.

However, I'm still encountering the same errors as before. Could you please explain how you reproduce this issue?

Errors
[auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: OperationProcessingError: unexpected JWT "iss" (issuer) claim value

**Pull request closed without merging**
